CMakeLists.txt

```diff
@@ -0,0 +1,162 @@
+if (BLAS_FOUND)
+    set(LIBS ${LIBS} ${BLAS_LIBRARIES})
+endif()
+
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+
+# common objects
+set(src ${PROJECT_SOURCE_DIR}/src/comms.F90
+    ${PROJECT_SOURCE_DIR}/src/constants.F90
+    ${PROJECT_SOURCE_DIR}/src/error_base.F90
+    ${PROJECT_SOURCE_DIR}/src/error.F90
+    ${PROJECT_SOURCE_DIR}/src/io.F90
+    ${PROJECT_SOURCE_DIR}/src/kmesh.F90
+    ${PROJECT_SOURCE_DIR}/src/readwrite.F90
+    ${PROJECT_SOURCE_DIR}/src/types.F90
+    ${PROJECT_SOURCE_DIR}/src/utility.F90
+    ${PROJECT_SOURCE_DIR}/src/ws_distance.F90)
+
+# wannier90.x specific
+set(srcw90 ${PROJECT_SOURCE_DIR}/src/disentangle.F90
+    ${PROJECT_SOURCE_DIR}/src/hamiltonian.F90
+    ${PROJECT_SOURCE_DIR}/src/overlap.F90
+    ${PROJECT_SOURCE_DIR}/src/plot.F90
+    ${PROJECT_SOURCE_DIR}/src/sitesym.F90
+    ${PROJECT_SOURCE_DIR}/src/transport.F90
+    ${PROJECT_SOURCE_DIR}/src/wannier90_readwrite.F90
+    ${PROJECT_SOURCE_DIR}/src/wannier90_types.F90
+    ${PROJECT_SOURCE_DIR}/src/wannierise.F90)
+
+# postw90.x specific
+set(srcp90 ${PROJECT_SOURCE_DIR}/src/postw90/berry.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/boltzwann.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/dos.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/geninterp.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/get_oper.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/gyrotropic.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/kpath.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/kslice.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/postw90_common.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/postw90_readwrite.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/postw90_types.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/spin.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/wan_ham.F90)
+
+# library specific
+set(srclib ${PROJECT_SOURCE_DIR}/src/library_interface.F90)
+
+# library objects to avoid recompilation of common objects
+add_library(obj_common OBJECT ${src})
+add_library(obj_wann90 OBJECT ${srcw90} $<TARGET_OBJECTS:obj_common>)
+add_library(obj_post90 OBJECT ${srcp90} $<TARGET_OBJECTS:obj_common>)
+add_library(obj_lib OBJECT ${srclib} $<TARGET_OBJECTS:obj_common> $<TARGET_OBJECTS:obj_wann90> )
+
+# shared library
+add_library(dynamic_lib SHARED $<TARGET_OBJECTS:obj_lib> $<TARGET_OBJECTS:obj_common> $<TARGET_OBJECTS:obj_wann90>)
+set_property(TARGET dynamic_lib PROPERTY VERSION ${PROJECT_VERSION})
+#set(soserial "2") # version 2 of library, but wannier90 version 4.0.0, to messy
+#set_property(TARGET dynamic_lib PROPERTY SOVERSION ${soserial})
+set_property(TARGET dynamic_lib PROPERTY OUTPUT_NAME wannier90)
+install(TARGETS dynamic_lib LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR})
+
+# static library
+add_library(static_lib STATIC $<TARGET_OBJECTS:obj_lib> $<TARGET_OBJECTS:obj_common> $<TARGET_OBJECTS:obj_wann90>)
+set_property(TARGET static_lib PROPERTY VERSION ${PROJECT_VERSION})
+set_property(TARGET static_lib PROPERTY OUTPUT_NAME wannier90)
+install(TARGETS static_lib ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
```

Comment:
Merge these and use `WANNIER90_SHARED_LIBS` that sets `BUILD_SHARED_LIBS` accordingly. The common approach is to perform two configure/build/installs if you want to provide both static and shared libraries.